src/main/kotlin/org/elm/ide/highlight/ElmSyntaxHighlightAnnotator.kt

```diff
@@ -12,74 +12,73 @@ import org.elm.lang.core.psi.elements.*
 class ElmSyntaxHighlightAnnotator : Annotator {
 
     override fun annotate(element: PsiElement, holder: AnnotationHolder) {
+        val h = Highlighter(holder)
```

Comment:
@ajalt what do you think about this? I wanted to partially apply the holder so that we don't have to pass it around so much. But this function will be called for every element in the Psi tree, so I assume that I should be careful about unnecessary heap allocations.

I don't know much about reading Kotlin/JVM bytecode, but it looks good:
```
   L1
    LINENUMBER 15 L1
    ALOAD 2
    INVOKESTATIC org/elm/ide/highlight/Highlighter.constructor-impl (Lcom/intellij/lang/annotation/AnnotationHolder;)Lcom/intellij/lang/annotation/AnnotationHolder;
    ASTORE 3
   L2
    LINENUMBER 16 L2
    ALOAD 1
    ASTORE 4
   L3
    LINENUMBER 17 L3
    ALOAD 4
    INSTANCEOF org/elm/lang/core/psi/elements/ElmValueDeclaration
    IFEQ L4
   L5
    ALOAD 3
    ALOAD 1
    CHECKCAST org/elm/lang/core/psi/elements/ElmValueDeclaration
    INVOKESTATIC org/elm/ide/highlight/Highlighter.valueDeclaration-impl (Lcom/intellij/lang/annotation/AnnotationHolder;Lorg/elm/lang/core/psi/elements/ElmValueDeclaration;)V
    GOTO L6
   L4
    LINENUMBER 18 L4
    ALOAD 4
    INSTANCEOF org/elm/lang/core/psi/elements/ElmTypeAnnotation
    IFEQ L7
   L8
    ALOAD 3
    ALOAD 1
    CHECKCAST org/elm/lang/core/psi/elements/ElmTypeAnnotation
    INVOKESTATIC org/elm/ide/highlight/Highlighter.typeAnnotation-impl (Lcom/intellij/lang/annotation/AnnotationHolder;Lorg/elm/lang/core/psi/elements/ElmTypeAnnotation;)V
    GOTO L6
```

Am I interpreting this correctly?

Reply:
This is calling static functions with the annotation holder as the first argument, so no extra heap allocations. You could go one small step further to avoid the `h.`s and move this function into the `Hightlighter`, and call it like `Highlighter(holder).highlight(element)`. Then all the other `Highlighter` methods could be private.